File: 388-longest-absolute-file-path/388-longest-absolute-file-path.cpp
//
// Created by loki on 2022/4/21.
//

#include "../common.h"

// C++自己写分隔字符串函数
void SplitString(string& s, vector<string>& v, const string& c)
{
    string::size_type pos1,pos2;
    pos2 = s.find(c);
    pos1 = 0;
    while (pos2!=string::npos){
        v.push_back(s.substr(pos1,pos2-pos1));
        pos1 = pos2 + c.size();
        pos2 = s.find(c,pos1);
    }
    if (pos1 != s.length())
        v.push_back(s.substr(pos1));
}

int lengthLongestPath(string input)
{
    int res=0;
    unordered_map<int,int> d;
    d[-1]=0;
    vector<string> lines;
    SplitString(input,lines,"\n");
    for (string line:lines) {
        int depth = count(line.begin(),line.end(),'\t');
        d[depth] = d[depth-1] + line.size() - depth;
        if (std::string::npos != line.find('.')) {
            res = max(res, d[depth] + depth);
        }
    }
    return res;
}

File: 388-longest-absolute-file-path/388-longest-absolute-file-path.py
class Solution:
    '''
    1、用 depth_length_map 保留每层路径的长度， input.split('\n') 切分为每行分析每行长度与文件
    2、line.count('\t') 的个数来判断是第几层
    3、line.count('.') 的个数判断是否有文件，有文件获取当前最长路径值
    4、每层都要添加depth个 / ， 长度需要修改
    '''
    def lengthLongestPath(self, input: str) -> int:
        res = 0
        depth_length_map={-1:0}
        for line in input.split('\n'):
            depth = line.count('\t')
            #每行空格最后要被去掉(\t)
            depth_length_map[depth] = depth_length_map[depth-1] + len(line) - depth
            if line.count('.'):
                res = max(res,depth_length_map[depth]+depth)
        return res